File: Arrays/ShuffleArray.js
var Solution = function(nums) {
    this.originalArray = [...nums];


    this.currentArray = [...nums];
};


Solution.prototype.reset = function() {
    this.currentArray = [...this.originalArray];


    return this.currentArray;
};


Solution.prototype.shuffle = function() {
    const shuffledArray = [...this.currentArray];


    for (let i = shuffledArray.length - 1; i > 0; i--) {
        const randomIndex = Math.floor(Math.random() * (i + 1));


        [shuffledArray[i], shuffledArray[randomIndex]] = 
        [shuffledArray[randomIndex], shuffledArray[i]];
    }


    return shuffledArray;
};

File: Arrays/ShuffleArray.cpp
#include <vector>
#include <algorithm>
#include <random>
#include <ctime>


class Solution {
private:
    std::vector<int> originalArray;
    std::vector<int> currentArray;
    std::mt19937 generator;

public:
    Solution(std::vector<int> nums) {
        originalArray = nums;
        currentArray = nums;


        generator.seed(std::time(nullptr));
    }
    

    std::vector<int> reset() {
        currentArray = originalArray;


        return currentArray;
    }
    

    std::vector<int> shuffle() {
        std::vector<int> shuffledArray = currentArray;


        std::shuffle(shuffledArray.begin(), shuffledArray.end(), generator);


        return shuffledArray;
    }
};
